File: lib/shire/catalog.ex
defmodule Shire.Catalog do
  @moduledoc """
  Reads agent catalog from static YAML files in priv/catalog/.
  No GenServer needed — files are read on demand.
  """

  defmodule Agent do
    @moduledoc false

    @type t :: %__MODULE__{
            name: String.t(),
            display_name: String.t(),
            description: String.t(),
            category: String.t(),
            emoji: String.t(),
            tags: [String.t()],
            harness: String.t(),
            model: String.t(),
            system_prompt: String.t()
          }

    defstruct [
      :name,
      :display_name,
      :description,
      :category,
      :emoji,
      :harness,
      :model,
      :system_prompt,
      tags: []
    ]
  end

  @spec list_agents() :: [Agent.t()]
  def list_agents do
    catalog_dir()
    |> Path.join("agents/**/*.yaml")
    |> Path.wildcard()
    |> Enum.map(&load_agent/1)
    |> Enum.reject(&is_nil/1)
  end

  @spec list_agents(keyword()) :: [Agent.t()]
  def list_agents(category: category) do
    list_agents()
    |> Enum.filter(&(&1.category == category))
  end

  def list_agents(_opts), do: list_agents()

  @spec get_agent(String.t()) :: Agent.t() | nil
  def get_agent(name) when is_binary(name) do
    if String.contains?(name, ["..", "/", "\\"]) do
      nil
    else
      case Path.wildcard(Path.join(catalog_dir(), "agents/**/#{name}.yaml")) do
        [path | _] -> load_agent(path)
        [] -> nil
      end
    end
  end

  @spec list_categories() :: [map()]
  def list_categories do
    path = Path.join(catalog_dir(), "categories.yaml")

    case YamlElixir.read_from_file(path) do
      {:ok, categories} ->
        Enum.map(categories, fn cat ->
          %{
            id: cat["id"],
            name: cat["name"],
            description: cat["description"] || ""
          }
        end)

      {:error, _} ->
        []
    end
  end

  @spec search(String.t()) :: [Agent.t()]
  def search(query) do
    query_down = String.downcase(query)

    list_agents()
    |> Enum.filter(fn agent ->
      String.contains?(String.downcase(agent.display_name || ""), query_down) or
        String.contains?(String.downcase(agent.description || ""), query_down) or
        Enum.any?(agent.tags || [], &String.contains?(String.downcase(&1), query_down))
    end)
  end

  defp load_agent(path) do
    case YamlElixir.read_from_file(path) do
      {:ok, data} ->
        %Agent{
          name: data["name"],
          display_name: data["display_name"],
          description: data["description"] || "",
          category: data["category"] || "",
          emoji: data["emoji"] || "",
          tags: data["tags"] || [],
          harness: data["harness"] || "claude_code",
          model: data["model"] || "claude-sonnet-4-6",
          system_prompt: data["system_prompt"] || ""
        }

      {:error, _} ->
        nil
    end
  end

  defp catalog_dir do
    Application.get_env(:shire, :catalog_dir) || Application.app_dir(:shire, "priv/catalog")
  end
end

File: src/services/catalog.ts
import { readFile, readdir, access } from "fs/promises";
import { join, basename, dirname } from "path";
import { fileURLToPath } from "url";
import { safeYamlLoad } from "../utils/yaml";

const __dirname = dirname(fileURLToPath(import.meta.url));

export interface CatalogAgent {
  name: string;
  displayName: string;
  description: string;
  category: string;
  emoji?: string;
  tags: string[];
  harness: string;
  model?: string;
  systemPrompt?: string;
}

export interface CatalogCategory {
  id: string;
  name: string;
  description: string;
}

function catalogDir(): string {
  return join(__dirname, "..", "..", "catalog");
}

async function fileExists(path: string): Promise<boolean> {
  try {
    await access(path);
    return true;
  } catch {
    return false;
  }
}

export async function listCategories(): Promise<CatalogCategory[]> {
  const path = join(catalogDir(), "categories.yaml");
  if (!(await fileExists(path))) return [];
  const content = await readFile(path, "utf-8");
  const data = safeYamlLoad(content) as CatalogCategory[];
  return data ?? [];
}

export async function listAgents(category?: string): Promise<CatalogAgent[]> {
  const agentsDir = join(catalogDir(), "agents");
  if (!(await fileExists(agentsDir))) return [];

  const result: CatalogAgent[] = [];
  const entries = await readdir(agentsDir, { withFileTypes: true });
  const categories = entries.filter((d) => d.isDirectory());

  for (const cat of categories) {
    if (category && cat.name !== category) continue;
    const catDir = join(agentsDir, cat.name);
    const allFiles = await readdir(catDir);
    const files = allFiles.filter((f) => f.endsWith(".yaml"));

    for (const file of files) {
      const content = await readFile(join(catDir, file), "utf-8");
      const data = safeYamlLoad(content) as Record<string, unknown>;
      if (!data) continue;

      result.push({
        name: String(data.name ?? basename(file, ".yaml")),
        displayName: String(data.display_name ?? data.name ?? ""),
        description: String(data.description ?? ""),
        category: cat.name,
        emoji: data.emoji ? String(data.emoji) : undefined,
        tags: Array.isArray(data.tags) ? data.tags.map(String) : [],
        harness: String(data.harness ?? "claude_code"),
        model: data.model ? String(data.model) : undefined,
        systemPrompt: data.system_prompt ? String(data.system_prompt) : undefined,
      });
    }
  }

  return result;
}

export async function getAgent(name: string): Promise<CatalogAgent | null> {
  const allAgents = await listAgents();
  return allAgents.find((a) => a.name === name) ?? null;
}

export async function searchAgents(query: string): Promise<CatalogAgent[]> {
  const q = query.toLowerCase();
  return (await listAgents()).filter(
    (a) =>
      a.displayName.toLowerCase().includes(q) ||
      a.description.toLowerCase().includes(q) ||
      a.tags.some((t) => t.toLowerCase().includes(q)),
  );
}
